### DISCUSSION TITLE
unable to get status from the device using napalm driver

### DISCUSSION THREAD

**Discussion by muradsabbah** [netbox-community/netbox#6283] [Q&A] [answered]
Hello,
I am new with Netbox  and complete the installation then i can remote  access the UI interface(http://x.x.x.x/dcim/devices/)  with username/password  and  add devices  but when I check the   device status or LLDP-neighbors  with Napalm driver  I get  this error:
 {"detail":"Error connecting to the device at x.x.x.x: Authentication to device failed.\n\nCommon causes of this problem are:\n1. Invalid username and password\n2. Incorrect SSH-key file\n3. Connecting to the wrong device\n\nDevice settings: cisco_ios x.x.x.x:22\n\n\nAuthentication failed."}

- it is working when i am using on local host (127.0.0.1)
-  i can ssh this cisco 1921 router from the server terminal (ubuntu ) that the Netbox where installed
-  i add the cisco router user/password in the configuration.py  
     **# Credentials that NetBox will uses to authenticate to devices when connecting via NAPALM.
            NAPALM_USERNAME = 'murad'
            NAPALM_PASSWORD = '***********'
            
            # NAPALM timeout (in seconds). (Default: 30)
            NAPALM_TIMEOUT = 30
            
              # NAPALM optional arguments (see https://napalm.readthedocs.io/en/latest/support/#optional- arguments). Argum>
            # be provided as a dictionary.
            NAPALM_ARGS = {
                'secret': '***********'
            }**
   

** Ubuntu 20.04.2 LTS (GNU/Linux 5.4.0-70-generic x86_64)
**:~$ uname -a
       Linux murad 5.4.0-70-generic #78-Ubuntu SMP Fri Mar 19 13:29:52 UTC 2021 x86_64 x86_64 x86_64 
      GNU/Linux

**Python 3.8.2 | Django 3.1.7 | NetBox 2.10.8

did i miss any configuration ?

**Comment by devzeroo** [comment]
i'm having the same scenario and error. Any luck ? @muradsabbah

**Comment by DanSheps** [comment]
I am not seeing any issues with Napalm on 3.1.5.  It is likely an authentication setting in netbox that is tripping you up.  You could use Napalm and try to query in nbshell directly.

**Comment by bogdancordos** [comment]
Make sure that the IP you are using to ssh from the server terminal is the same IP as the Primary IP in NetBox for that device.

**Reply by devzeroo** [reply]
Solved here thanks!

**Reply by raviv0731** [reply]
Hello, i am facing similar issue with the same error. now, i am not sure about secret. Does the devices need secrets to generate those status/lldp nei/configuration tabs? Do we need to create one ? if so where ? i don't see any secrets in netbox. i read its now comes as a plugin etc. Do i have to install it ?

**Comment by devzeroo** [comment, marked as answer]
Just had to add the secret in configuration.py NAPALM parameters and also enable XML permission on my Tacacs.

**Reply by raviv0731** [reply]
Hello, i am facing similar issue with the same error. now, i am not sure about secret. Does the devices need secrets to generate those status/lldp nei/configuration tabs?  Do we need to create one ? if so where ? i don't see any secrets in netbox. i read  its now comes as a plugin etc. Do i have to install it ?

Any help is greatly appreciated. Thanks.

**Reply by candlerb** [reply]
"secret" in this context is a Cisco term, it's the extra password used after login with "enable" to get into privileged mode.  It's not a Netbox concept as such.  You set it using `NAPALM_ARGS={"secret":"foo"}`

The Netbox secrets plugin is nothing to do with this.

**Reply by raviv0731** [reply]
Thanks for the clarification **Candlerb**. However in that case i have updated the secret as well. also on the other note, i also removed username/password on config file and did docker-compose down/up but no actual vm reboot.  to just update them on "config revisions". However they are still graded out for me. I am unsure if that would resolve the issue. Any more hints on this please?

**Reply by raviv0731** [reply]
The below script works just fine. it pulls the config. However through Netbox we see the above error. 

**import napalm
import json
import sys
import os

driver = napalm.get_network_driver ('ios')

device = driver(hostname ='172.20.177.2',username='xxxx',password='xxxxxx')
device.open()
print(device.get_config(retrieve='running', full=True))
print(device.get_environment())
device.close()**
